webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/repository/ldap/LDAPEntityMapper.scala

```diff
@@ -532,34 +528,20 @@ class LDAPEntityMapper(
 
   // two utilities to serialize / deserialize Map[TechniqueVersion,DateTime]
   def unserializeAcceptations(value: String): PureResult[Map[TechniqueVersion, DateTime]] = {
-    import net.liftweb.json.JsonAST.JField
-    import net.liftweb.json.JsonAST.JString
-    import net.liftweb.json.JsonParser.*
-
-    parse(value) match {
-      case JObject(fields) =>
-        fields.collect { case JField(version, JString(date)) => (version, date) }.traverse {
-          case (v, d) =>
-            TechniqueVersion
-              .parse(v)
-              .leftMap(Unexpected.apply)
-              .flatMap(version => {
-                try {
-                  Right((version, GeneralizedTime(d).dateTime))
-                } catch {
-                  case NonFatal(ex) => Left(SystemError(s"Error when trying to parse '${d}' as a datetime", ex))
-                }
-              })
-        }.map(_.toMap)
-      case _               => Right(Map())
-    }
+    (for {
+      l <- value.fromJson[Map[String, String]]
+      t <- l.toList.traverse {
+             case (a, b) =>
+               for {
+                 tv <- TechniqueVersion.parse(a)
+                 dt <- GeneralizedTime.parse(b).toRight(s"Error when parsing '${b}' as a generalized time'")
+               } yield (tv, dt.dateTime)
+           }
+    } yield t.toMap).left.map(s => Unexpected(s))
```

Comment:
I didn't do it because I'm really not sure it worth it. 

There is several saillant differences from standard object that make it less nice, almost bad: 
- there is no standard place to put the codec. Contrary to a case class where it lives in the the companion object and it's the standard place to look at, here, we need to put it somewhere and when we need it, find the correct class where it was defined ; (see at the end of `LDAPEntityMapper`, its companion object)
- the quantity of plumbing is more than the direct implementation, 
- we need to encapsulate the `DateTime` codec in a sub-object to avoid importing them elsewhere because of the `import LDAPEntityMapper.*`
- but more dangerously, for a newcomer, looking for decoding `Map[TechniqueVersion, DateTime]`, the use will very likely look for `TechniqueVersion` decoder, find it, a `DateTime` decoder, find it, and map the result to map. Especially someone who doesn't know `zio-json` special case for map keys. And the result will be an incorrect decoding. 

All that let me think that: the root problem is that decoding `Map[A, B]` by themselve is not idiomatic at all in `zio-json`. Either you decode the whole englobing object and in the that object companion, you put the specific map codec, or you encapsulate the Map into an other object and make the decoder in its companion (circling back to previous case), 

I wanted to avoid changing the API of active technique, so I also avoided going to the full "encapsulate map into a `AcceptationTime` object" even if it's clearly the correct, principled solution, because it was not the goal of that refactoring. 

But since I wrote all the plumbing to show you the problems, I will do the last miles in a subsequent commit.

Reply:
I agree that it may be significantly harder to make something out of all the declarative values, than just a "mapper" class, but the boilerplate is reduced to almost the same cost I think specifically when creating the "opaque type" like you did (we only need to import zio.json.* instead of instantiating the mapper).
Many lib also use the same declarative style so it would not be hard to migrate (I hope we won't need to do that !)